src/core/src/local_loggers/CompositeLogger.py

```diff
@@ -22,20 +22,26 @@
 class CompositeLogger(object):
     """ Manages diverting different kinds of output to the right sinks for them. """
 
-    def __init__(self, env_layer=None, file_logger=None, current_env=None):
+    def __init__(self, env_layer=None, file_logger=None, telemetry_writer=None, current_env=None):
         self.env_layer = env_layer
         self.file_logger = file_logger
+        self.telemetry_writer = telemetry_writer
         self.ERROR = "ERROR:"
         self.WARNING = "WARNING:"
         self.DEBUG = "DEBUG:"
         self.VERBOSE = "VERBOSE:"
         self.current_env = current_env
         self.NEWLINE_REPLACE_CHAR = " "
+        self.telemetry_task = "ExtensionCoreLog"
+        self.telemetry_msg_truncated_note = " This message will be truncated in telemetry as it exceeds the size limit."
 
-    def log(self, message):
+    def log(self, message, message_type=Constants.TelemetryEventLevel.Informational):
         """log output"""
         message = self.__remove_substring_from_message(message, Constants.ERROR_ADDED_TO_STATUS)
         message = message.strip()
+        if self.telemetry_writer is not None and self.telemetry_writer.events_folder_path is not None:
+            self.telemetry_writer.write_event(self.telemetry_task, message, message_type)
+            message = self.__add_telemetry_error_note(message)
```

Comment:
Telemetry is not best effort.. The design tenet here is:
- Does the agent support telemetry? No = don't allow execution to proceed. Write the error to CRP through the status file asking the customer to update their agent.
- If the agent supports telemetry, but telemetry is somehow not working (in a way we can detect), we need to stop execution again with the same process.

The only exception to this would be to allow developer validation at our end.